kernels/portable/cpu/op_grid_sampler_2d.cpp

```diff
@@ -0,0 +1,257 @@
+//  Provided subject to the LICENSE file in the top level directory.
+//
+
+#include <executorch/kernels/portable/cpu/scalar_utils.h>
+#include <executorch/kernels/portable/cpu/util/elementwise_util.h>
+#include <executorch/runtime/kernel/kernel_includes.h>
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <utility>
+
+#include <c10/util/irange.h>
+
+namespace torch {
+namespace executor {
+namespace native {
+
+using Tensor = exec_aten::Tensor;
+using ScalarType = executorch::aten::ScalarType;
+using SizesType = executorch::aten::SizesType;
+
+// Transform normalized coordinates to pixel space
+inline float unnormalize_coord(float coord, int64_t size, bool align_corners) {
+  if (align_corners) {
+    // -1 and 1 correspond to the centers of the first and last pixels
+    return ((coord + 1.0f) / 2.0f) * (size - 1);
+  } else {
+    // -1 and 1 correspond to the boundary of the image
+    return ((coord + 1.0f) * size - 1.0f) / 2.0f;
+  }
+}
+
+// Compute source index and interpolation weight
+inline std::pair<int64_t, float>
+compute_source_index_and_weight(float coord, int64_t size, bool align_corners) {
+  float real_coord = unnormalize_coord(coord, size, align_corners);
+  int64_t index = std::floor(real_coord);
+  float weight = real_coord - index;
+  return {index, weight};
+}
+
+// Apply reflective padding to handle out-of-bounds coordinates
+inline int64_t reflect_coord(int64_t coord, int64_t size) {
+  if (size <= 1)
+    return 0;
+
+  int64_t double_size = 2 * size - 2;
+  if (double_size <= 0)
+    return 0;
+
+  // Handle negative coordinates
+  int64_t abs_coord = std::abs(coord);
+  abs_coord = abs_coord % double_size;
+  if (abs_coord >= size) {
+    abs_coord = double_size - abs_coord;
+  }
+
+  return abs_coord;
+}
+
+// Get pixel value with proper boundary handling based on padding mode
+template <typename T>
+T get_pixel_value(
+    const T* input_data,
+    int64_t n,
+    int64_t c,
+    int64_t h,
+    int64_t w,
+    int64_t height,
+    int64_t width,
+    int64_t padding_mode) {
+  // Handle out-of-bounds coordinates
+  if (h < 0 || h >= height || w < 0 || w >= width) {
+    if (padding_mode == 0) { // Zeros
+      return 0;
+    } else if (padding_mode == 1) { // Border
+      h = std::min(std::max(h, static_cast<int64_t>(0)), height - 1);
+      w = std::min(std::max(w, static_cast<int64_t>(0)), width - 1);
+    } else if (padding_mode == 2) { // Reflection
+      h = reflect_coord(h, height);
+      w = reflect_coord(w, width);
+    }
+  }
+
+  // Calculate offset using strides for memory access
+  const int64_t batch_stride = c * height * width;
+  const int64_t channel_stride = height * width;
+  const int64_t height_stride = width;
+
+  return input_data
+      [n * batch_stride + c * channel_stride + h * height_stride + w];
+}
+
+// Process grid sampling with specified input, grid, and modes
+template <typename T>
+void grid_sampler_2d_impl(
```

Comment:
This implementation requires default dim order. Add a check in your check_args function, to check that the dim order is the default.